File: hse/BPM161/abramov/2/LockFreeSet/src/test/java/ru/spbhse/karvozavr/lockfreeset/LockFreeSetTest.java
package ru.spbhse.karvozavr.lockfreeset;

import org.junit.Test;

import static org.junit.Assert.*;

public class LockFreeSetTest {

    @Test
    public void testAddRemove() {
        LockFreeSet<Integer> set = new LockFreeSet<>();

        for (int i = 0; i < 10; ++i) {
            set.add(i);
        }

        for (int i = 0; i < 10; ++i) {
            assert set.contains(i);
        }

        for (int i = 0; i < 10; ++i) {
            set.remove(i);
        }

        for (int i = 0; i < 10; ++i) {
            assertFalse(set.contains(i));
        }
    }
}

File: hse/BPM161/vasilenko/2/src/test/kotlin/lockfree/LockFreeSetTest.kt
package lockfree

import org.hamcrest.Matchers.containsInAnyOrder
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class LockFreeSetTest {

    private lateinit var set : LockFreeSet<Int>

    @Before
    fun setUp() {
        set = LockFreeSet()
    }

    @Test
    fun isEmpty() {
        assertTrue(set.isEmpty)
        fill(3)
        assertFalse(set.isEmpty)
        clear(3)
        assertTrue(set.isEmpty)
    }

    private fun clear(n: Int) {
        for (i in 1..n)
            set.remove(i)
    }

    @Test
    fun add() {
        assertTrue(set.add(1))
        assertFalse(set.add(1))
        assertTrue(set.add(2))
        assertFalse(set.add(2))
        assertTrue(set.add(3))
        assertFalse(set.add(1))
        assertFalse(set.add(3))
        assertFalse(set.add(2))
    }

    @Test
    fun remove() {
        repeat(2) {
            for (i in 1..3)
                assertFalse(set.remove(i))
            fill(3)
            for (i in 1..3)
                assertTrue(set.remove(i))
        }
    }

    private fun fill(n: Int) {
        for (i in 1..n)
            set.add(i)
    }

    @Test
    fun contains() {
        repeat(2) {
            for (i in 1..3)
                assertFalse(set.contains(i))
            fill(3)
            for (i in 1..3)
                assertTrue(set.contains(i))
            clear(3)
        }
    }

    @Test
    fun iterator() {
        assertFalse(set.iterator().hasNext())
        fill(3)
        assertThat(set.iterator().asSequence().toList(), containsInAnyOrder(1, 2, 3))
        clear(1)
        assertThat(set.iterator().asSequence().toList(), containsInAnyOrder(2, 3))
        clear(3)
        assertFalse(set.iterator().hasNext())
    }
}